### DISCUSSION TITLE
[@types/node] NodeJS.ReadableStream is not a generic

### DISCUSSION THREAD

**Discussion by agirorn** [DefinitelyTyped/DefinitelyTyped#59532] [Issues with a @types package]
I think the NodeJS.ReadableStream should be generic since it can be created as an object stream and the objects could be typed if it were generic.

When I try to do use it as a generic like this I get " Type 'ReadableStream' is not generic."

```ts
  interface MyType {
    messageID: string:
  }

  async load(events: NodeJS.ReadableStream<MyType>) {
    for await (const event of events) {
      console.log(event);
    }
  }
```

[a comment by an automated account is not shown]

**Comment by SimonSchick** [comment]
Unfortunately this is not a straight forward task due to the often specifialized nature of streams using buffers and the ecosystem, we may see this happen at some point.

**Reply by agirorn** [reply]
Maybe I'm missing something but the current typing uses some specific type for the stream I think it is `string | any` so why could that not be generic, something like this?

```typescript
interface ReadableStream<T=string | any> {
  ...
  // Using T
}
```

Then all the legacy implementation would work but custom streams could be created where the streams types had better typed.

**Reply by trevorah** [reply]
I've got a pull request (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60979) that adds generics to readable streams in a similar way. I'm attempting to maintaining backwards compatibility by defining the type as `ReadableStream<R = any>` where binary mode streams are defined like `ReadableStream<string | Buffer>` to keep with the node buffer ecosystem.

**Comment by leftshift** [comment]
Seeing all progress on this getting stalled is quite disappointing, as the current state makes it close to impossible to provide an at least somewhat typesafe public api that accepts ReadableStreams in object mode. :/

I don't quite see how 'this has some risk of breaking existing code' (which funnily enough seems to still include [a funny little wrong type](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/60526), which might break stuff at runtime, much worse imho than producing new compile-time errors) outweighs the benefits of providing a way to use core features of a part of nodejs in a typesafe manner.
